### ISSUE TITLE
Error on using Counter's __iadd__ and __isub__ with dict[Any, int]

### ISSUE THREAD

**Issue by tyilo** [python/mypy#15288]
**Bug Report**

`Counter` supports updating with `+=` and `-=` using a `dict[Any, int]`, however mypy disallows this.

**To Reproduce**

```python
from collections import Counter

ctr: Counter[str] = Counter()
ctr += {"a": 2}
ctr -= {"a": 1}

print(ctr)
```

https://mypy-play.net/?mypy=master&python=3.11&gist=3b6ad9610e9a3d59d2838ae26c47903a

**Expected Behavior**

mypy shouldn't report any errors.

**Actual Behavior**

```
main.py:4: error: Argument 1 to "__iadd__" of "Counter" has incompatible type "dict[str, int]"; expected "Counter[str]"  [arg-type]
main.py:5: error: Argument 1 to "__isub__" of "Counter" has incompatible type "dict[str, int]"; expected "Counter[str]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: master branch
- Python version used: 3.11
